### DISCUSSION TITLE
Negative time ranges

### DISCUSSION THREAD

**Discussion by jotaen** [jotaen/klog#109] [Feature ideas & discussions]
In analogy to tracking a negative duration (`-1h`) it could be useful to allow tracking “negative” time ranges, i.e. a specific period of time to be deduced from the total.

Syntax ideas:

```
2020-01-01
    8:00-17:00 Work
    -12:00-13:00 Lunch break
    -(12:00-13:00) Lunch break
    -[12:00-13:00] Lunch break
```

Currently favouring the `[]` syntax, as it’s more distinct from the should-total.

Source: https://github.com/jotaen/klog/issues/3

**Comment by acikek** [comment]
Why does it need to be wrapped in `()` or `[]`?

**Comment by jotaen** [comment]
My reasoning was for visual clarity. It’s not necessary from a technical point of view, but I feel that it might be too easy to overlook otherwise. I added the “naked” version as well in the example above, I’m also open to any other suggestion.

**Comment by vladdeSV** [comment]
I am in favor of wrapping in `()`, with the same reasoning as @jotaen:

> for visual clarity.

**Comment by vladdeSV** [comment]
How would whitespace be handled inside and before the parentheses?

Are the following okay?
```klog
2021-04-20
  -(0:00-1:00) foo
  -    ( 0:00   -      1:00  ) foo
```

**Comment by ulken** [comment]
I would love to see this happen. Personally, no extra syntax is needed as the syntax highlighting would make it obvious, but maybe people don't use that?

If still deemed necessary, `()` resembles negating a mathematical expression, so that's makes sense. On the other hand, `[]` indicates the endpoints are to be included in the interval (see [Notation for intervals](https://en.wikipedia.org/wiki/Interval_(mathematics)#Notations_for_intervals)), which is the case here?

**Comment by jotaen** [comment]
Cross-posting from https://github.com/jotaen/klog/discussions/244#discussioncomment-4882157, regarding the current state of affairs here:

Negative time ranges are not off the table. Realistically speaking, however, I have to say that adding negative time ranges is a pretty comprehensive undertaking – deciding on the syntax and semantics is the easiest part, but extending the spec, writing and refactoring the code, and documenting it will be substantial effort. It might eventually happen, but it would certainly take a while.

**Reply by jotaen** [reply]
The implementation is fairly comprehensive, and the spec also needs changing – so I don’t know when I’ll get around to starting it, let alone when it’s done. Here are some considerations: (see also https://github.com/jotaen/klog/issues/252.)

# File format

## Notation

I think what’s mostly clear is that the negative time range needs to be preceded by a minus sign `-`. There should also be a clear visual distinction in regards to regular time ranges. The following options are on the table right now:

```
2000-01-01
  -(15:00-16:00)
  -[15:00-16:00]
```

- Parentheses `()`:
  - That’s both common as mathematical notation but also in daily writing, so it seems intuitive.
  - Parenthesis are already used for should totals `(8h!)`, so it might create ambiguity – but maybe not since the should total is in a different place and obviously serves a different purpose.
- Square brackets `[]`:
  - Also used in math as notation for inclusive intervals (which would be the case here).
  - Might look more “technical”, as square brackets are less common than parentheses in daily writing.

In the end, there should be only one of these notations.

## Open negative time range

Regular time ranges can be open (`15:00-?`), so it would make sense for negative time ranges to be able to be open as well.

Records are currently constrained to one open time range at most. Likewise, there should be just one open negative time range. But it would make sense for both to allow that independent of each other, to support something like:

```
2000-01-01
  8:00-? Work
  -(12:00-?) Lunch break
```

Otherwise, you would have to close the `Work` entry before starting a break, and you probably don’t want that.

## Whitespace

Time ranges can have spaces around the dash (`15:00-16:00` vs `15:00 - 16:00`, both fine). Negative time ranges can have that as well. But what about:

- `-   (15:00 - 16:00)`
  - I’m tending to think this shouldn’t be allowed, as you could confuse it with a bullet point. I also don’t see any good reason **for** it.
- `-(   15:00 - 16:00      )`
  - Maybe people prefer some padding inside the parenthesis? Maybe nothing wrong with this?

# CLI tool

## `klog pause`

The `klog pause` command currently opens a negative duration and keeps decrementing that. Once negative time ranges are there, `klog pause` should probably default to starting an open negative time range, so the semantics could be changed as follows:

- `klog pause` would default to append a new negative open range to the record. (I.e., the counterpart of `klog start`.)
- `klog pause --follow` (`klog pause -f`) would do the same, but keep the shell process open. Once you hit `^C`, the open entry is closed.
- `klog pause --duration` (`klog pause -d`) would be the “old” behaviour. (`--follow`/`-f` would be implicit here, so you can specify it as well, but also omit it, for convenience.)

There also would need to be a new command for finishing the negative open range, e.g. `klog unpause` or `klog resume`. (I.e., the counterpart of `klog stop`.)
